File: 1365. How Many Numbers Are Smaller Than the Current Number.java
// Easy
// Array
// Time:O(nlogn), Space:O(n)
// https://leetcode.cn/problems/how-many-numbers-are-smaller-than-the-current-number/

// Actually I can't find any (looks smarter) way to do this....

import java.util.Arrays;

class Solution {
    public int[] smallerNumbersThanCurrent(int[] nums) {
        int n = nums.length;
        int[][] cnt = new int[n][2];

        for (int i = 0; i < n; i++) {
            cnt[i][0] = nums[i];
            cnt[i][1] = i;
        }

        Arrays.sort(cnt, (a, b) -> a[0] - b[0]);

        int[] res = new int[n];
        for (int i = 0; i < n; i++) {
            if (i > 0 && cnt[i][0] == cnt[i - 1][0]) {
                res[cnt[i][1]] = res[cnt[i - 1][1]];
            } else {
                res[cnt[i][1]] = i; // i' smaller than it
            }
        }
        return res;
    }
}

File: 1001-2000/1365. How Many Numbers Are Smaller Than the Current Number.py
# Time:O(nlogn), Space:O(n)

from typing import List

class Solution:
    def smallerNumbersThanCurrent(self, nums: List[int]) -> List[int]:
        sorted_num = sorted(nums)

        first_idx = {}
        for i in range(len(sorted_num)):
            if sorted_num[i] not in first_idx:
                first_idx[sorted_num[i]] = i

        res = []
        for num in nums:
            res.append(first_idx[num])

        return res
